source/slang/slang-parser.cpp

```diff
@@ -5579,6 +5589,119 @@ namespace Slang
         return varDeclrStatement;
     }
 
+    static Expr* constructIfLetPredicate(Parser* parser, VarExpr* varExpr)
+    {
+        // create a "var.hasValue" expression
+        MemberExpr* memberExpr = parser->astBuilder->create<MemberExpr>();
+        memberExpr->baseExpression = varExpr;
+        parser->FillPosition(memberExpr);
+        memberExpr->name = getName(parser, "hasValue");
+
+        // create a "==" operator
+        auto opExpr = parser->astBuilder->create<VarExpr>();
+        opExpr->name = getName(parser, "==");
+        opExpr->scope = parser->currentScope;
+        parser->FillPosition(opExpr);
+
+        // create a 'true' literal
+        BoolLiteralExpr* boolLiteralExpr = parser->astBuilder->create<BoolLiteralExpr>();
+        boolLiteralExpr->value = true;
+
+        // create a "var.hasValue == true" expression
+        InfixExpr* infixExpr = parser->astBuilder->create<InfixExpr>();
+        infixExpr->loc = opExpr->loc;
+        infixExpr->functionExpr = opExpr;
+        infixExpr->arguments.add(memberExpr);
+        infixExpr->arguments.add(boolLiteralExpr);
+        return infixExpr;
+    }
+
+    // Parse the syntax 'if (let var = X as Y)'
+    Stmt* Parser::parseIfLetStatement()
+    {
+        ScopeDecl* scopeDecl = astBuilder->create<ScopeDecl>();
+        pushScopeAndSetParent(scopeDecl);
+
+        SeqStmt* newBody = astBuilder->create<SeqStmt>();
+
+        IfStmt* ifStatement = astBuilder->create<IfStmt>();
+        FillPosition(ifStatement);
+        ReadToken("if");
+        ReadToken(TokenType::LParent);
+
+        // parse 'let var = X as Y'
+        ReadToken("let");
+        auto identifierToken = ReadToken(TokenType::Identifier);
+        ReadToken(TokenType::OpAssign);
+        auto initExpr = ParseInitExpr();
+
+        // insert 'let tempVarDecl = X as Y;'
+        auto tempVarDecl = astBuilder->create<LetDecl>();
+        tempVarDecl->nameAndLoc = NameLoc(getName(this, "OptVar"), identifierToken.loc);
```

Comment:
Maybe "$optVar" to avoid conflicting with names in user code.

Reply:
done